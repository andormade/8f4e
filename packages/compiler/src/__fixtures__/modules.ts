export default [
	{
		state: {},
		code: `
            module abs
        
            # memory
            memory DEFAULT_VALUE 0
            memory in &DEFAULT_VALUE
            memory out 0
        
            # locals
            local input
        
            # code
            push &out
            push in
            localSet input
            push input
            push 0
            lessThan
            if 
                push 0
                push input
                sub
            else
                push input
            end
            store
        `.split('\n'),
	},
];
