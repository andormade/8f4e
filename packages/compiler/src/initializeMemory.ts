import { CompiledModule, Connection, MemoryAddressLookup, MemoryBuffer } from './types';

export function generatememoryAddressLookup(compiledModules: CompiledModule[]): MemoryAddressLookup {
	const lookup = {};
	compiledModules.forEach(({ offset, memoryAddresses, moduleId }) => {
		lookup[moduleId] = offset;
		memoryAddresses.forEach(({ id, address }) => {
			lookup[moduleId + '_' + id] = address;
		});
	});
	return lookup;
}

export function setUpConnections(
	memoryBuffer: MemoryBuffer,
	memoryAddresses: MemoryAddressLookup,
	connections: Connection[]
): void {
	connections.forEach(connection => {
		const inputName = connection.toModuleId + '_' + connection.toConnectorId;
		const outputName = connection.fromModuleId + '_' + connection.fromConnectorId;
		const inputAddress = memoryAddresses[inputName];
		const outputAddress = memoryAddresses[outputName];
		memoryBuffer[inputAddress / memoryBuffer.BYTES_PER_ELEMENT] = outputAddress;
	});
}
