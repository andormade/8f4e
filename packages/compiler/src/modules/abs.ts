import { ModuleGenerator } from '../types';
import { compile } from 'module-compiler';

const abs: ModuleGenerator<unknown> = function (moduleId, offset) {
	return compile(
		`
		# abs module
		
		# memory
		private DEFAULT_VALUE 0
		inputPointer in DEFAULT_VALUE
		output out 0
		
		# locals
		local input
		
		# code
		const out
		const in
		load
		load
		localSet input
		localGet input
		const 0
		lessThan
		if 
			const 0
			localGet input
			sub
		else
			localGet input
		end
		store
	`,
		moduleId,
		offset.byte(0)
	);
};

export default abs;
