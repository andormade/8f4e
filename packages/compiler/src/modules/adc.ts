import {
	Instruction,
	Type,
	createFunctionBody,
	createLocalDeclaration,
	i32const,
	i32load,
	i32store,
	ifelse,
	localGet,
	localSet,
} from '@8f4e/bytecode-utils';
import { MemoryMap, MemoryTypes, ModuleGenerator, Output } from '../types';
import { I16_SIGNED_LARGEST_NUMBER, LOGIC_HIGH, LOGIC_LOW } from '../consts';

export enum Memory {
	DEFAULT_VALUE,
	INPUT_POINTER,
	OUTPUT_1,
	OUTPUT_2,
	OUTPUT_3,
	OUTPUT_4,
	OUTPUT_5,
	OUTPUT_6,
	OUTPUT_7,
	OUTPUT_8,
	OUTPUT_9,
	OUTPUT_10,
	OUTPUT_11,
	OUTPUT_12,
	OUTPUT_13,
	OUTPUT_14,
	OUTPUT_15,
	OUTPUT_16,
}

const masks = [
	[Memory.OUTPUT_1, 0b0000000000000001],
	[Memory.OUTPUT_2, 0b0000000000000010],
	[Memory.OUTPUT_3, 0b0000000000000100],
	[Memory.OUTPUT_4, 0b0000000000001000],
	[Memory.OUTPUT_5, 0b0000000000010000],
	[Memory.OUTPUT_6, 0b0000000000100000],
	[Memory.OUTPUT_7, 0b0000000001000000],
	[Memory.OUTPUT_8, 0b0000000010000000],
	[Memory.OUTPUT_9, 0b0000000100000000],
	[Memory.OUTPUT_10, 0b0000001000000000],
	[Memory.OUTPUT_11, 0b0000010000000000],
	[Memory.OUTPUT_12, 0b0000100000000000],
	[Memory.OUTPUT_13, 0b0001000000000000],
	[Memory.OUTPUT_14, 0b0010000000000000],
	[Memory.OUTPUT_15, 0b0100000000000000],
	[Memory.OUTPUT_16, 0b1000000000000000],
];

enum Locals {
	INPUT,
	__LENGTH,
}

export interface Config {
	resolution?: number;
}

const adc: ModuleGenerator<Config> = function (moduleId, offset, { resolution = 8 } = {}) {
	const functionBody = createFunctionBody(
		[createLocalDeclaration(Type.I32, Locals.__LENGTH)],
		[
			...i32const(offset.byte(Memory.INPUT_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(resolution === 16 ? 1 : Math.floor(I16_SIGNED_LARGEST_NUMBER / (Math.pow(2, resolution) - 1))),
			Instruction.I32_DIV_S,
			...localSet(Locals.INPUT),

			...masks
				.slice(0, resolution)
				.map(([memoryAddress, mask]) => [
					...i32const(offset.byte(memoryAddress)),
					...localGet(Locals.INPUT),
					...i32const(mask),
					Instruction.I32_AND,
					...i32const(0),
					Instruction.I32_GT_S,
					...ifelse(Type.I32, i32const(LOGIC_HIGH), i32const(LOGIC_LOW)),
					...i32store(),
				])
				.flat(),
		]
	);

	return {
		moduleId,
		functionBody,
		byteAddress: offset.byte(0),
		wordAddress: offset.word(0),
		memoryMap: [
			{
				type: MemoryTypes.PRIVATE,
				address: Memory.DEFAULT_VALUE,
				default: 0,
			},
			{
				type: MemoryTypes.INPUT_POINTER,
				address: Memory.INPUT_POINTER,
				id: 'in',
				default: offset.byte(Memory.DEFAULT_VALUE),
			},
			...(masks.slice(0, resolution).map(([memoryAddress], index) => ({
				type: MemoryTypes.OUTPUT,
				address: memoryAddress,
				id: 'out:' + (index + 1),
				default: 0,
			})) as Output[]),
		],
	};
};

export default adc;
