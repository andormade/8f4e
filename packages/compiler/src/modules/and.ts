import { i32load, i32const, i32store, ifelse, br } from '../wasm/instructions';
import { createFunctionBody } from '../wasm/sections';
import { Instruction, Type } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';
import { I16_SIGNED_LARGEST_NUMBER } from '../consts';

export const enum Memory {
	DEFAULT_VALUE,
	INPUT_1_POINTER,
	INPUT_2_POINTER,
	OUTPUT,
}

const and: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(offset(Memory.INPUT_1_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(0),
			Instruction.I32_GT_S,
			...ifelse(Type.VOID, [
				...i32const(offset(Memory.INPUT_2_POINTER)),
				...i32load(),
				...i32load(),
				...i32const(0),
				Instruction.I32_GT_S,
				...ifelse(Type.VOID, [
					...i32const(offset(Memory.OUTPUT)),
					...i32const(I16_SIGNED_LARGEST_NUMBER),
					...i32store(),
					...br(2),
				]),
			]),

			...i32const(offset(Memory.OUTPUT)),
			...i32const(0),
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset: offset(0),
		initialMemory: [0, offset(Memory.DEFAULT_VALUE), offset(Memory.DEFAULT_VALUE), 0],
		memoryAddresses: [
			{ address: offset(Memory.OUTPUT), id: 'out' },
			{
				address: offset(Memory.INPUT_1_POINTER),
				id: 'in:1',
				default: offset(Memory.DEFAULT_VALUE),
			},
			{
				address: offset(Memory.INPUT_2_POINTER),
				id: 'in:2',
				default: offset(Memory.DEFAULT_VALUE),
			},
		],
	};
};

export default and;
