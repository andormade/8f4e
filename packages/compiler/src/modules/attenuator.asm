private zero 0
inputPointer in zero
private divisor 1
output out 0

const out
const in
load
load
const divisor
load
div
store