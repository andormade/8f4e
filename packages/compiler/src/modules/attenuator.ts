import { i32const, i32load, i32store } from '../wasm/instructions';
import { createFunctionBody } from '../wasm/sections';
import { Instruction } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';

const enum Memory {
	ZERO,
	INPUT_POINTER,
	DIVISOR,
	OUT,
}

const attenuator: ModuleGenerator = function (moduleId, offset, initialConfig) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(offset(Memory.OUT)),
			...i32const(offset(Memory.INPUT_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(offset(Memory.DIVISOR)),
			...i32load(),
			Instruction.I32_DIV_S,
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset: offset(0),
		initialMemory: [0, offset(Memory.ZERO), initialConfig.divisor, 0],
		memoryAddresses: [
			{ address: offset(Memory.OUT), id: 'out' },
			{ address: offset(Memory.DIVISOR), id: 'divisor' },
			{ address: offset(Memory.INPUT_POINTER), id: 'in', isInputPointer: true },
		],
	};
};

export default attenuator;
