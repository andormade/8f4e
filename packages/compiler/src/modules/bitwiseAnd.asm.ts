export default `
	private default1 0
	private default2 0
	inputPointer in:1 default1
	inputPointer in:2 default2
	output out 0
		
	const out
	load in:1
	load
	load in:2
	load
	and
	store
`;
