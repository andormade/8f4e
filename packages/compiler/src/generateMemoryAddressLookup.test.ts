import modules from './__fixtures__/modules';
import { generateMemoryAddressLookup } from './initializeMemory';

import { compileModules } from '.';

describe('compiler', () => {
	test('generateMemoryAddressLookup', () => {
		const compiledModules = compileModules(modules, { startingMemoryWordAddress: 0, constants: {} });
		expect(generateMemoryAddressLookup(compiledModules)).toMatchSnapshot();
	});
});
