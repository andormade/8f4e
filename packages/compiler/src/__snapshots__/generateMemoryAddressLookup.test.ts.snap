// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler generateMemoryAddressLookup 1`] = `
Map {
  "abs__startAddress" => 0,
  "absDEFAULT_VALUE" => 0,
  "absin" => 1,
  "absout" => 2,
}
`;
